File: frontend/src/types/document.ts
export type SourceType = "pdf";
export type EditorMode = "view" | "quick-edit" | "document-edit";

export interface ElementSource {
  pageNumber: number;
  originalText?: string;
  originalX?: number;
  originalY?: number;
  originalWidth?: number;
  originalHeight?: number;
  isNew: boolean;
}

export interface TextElement {
  id: string;
  type: "text";
  content: string;
  x: number;
  y: number;
  width: number;
  height: number;
  fontSize: number;
  fontFamily: string;
  fontWeight: "normal" | "bold";
  fontStyle?: "normal" | "italic";
  color: string;
  alignment: "left" | "center" | "right";
  rotation: number;
  source: ElementSource;
}

export type DocumentElement = TextElement;

export interface PageModel {
  pageNumber: number;
  width: number;
  height: number;
  elements: DocumentElement[];
}

export interface DocumentModel {
  sourceType: SourceType;
  fileName: string;
  pageCount: number;
  pages: PageModel[];
}

export interface SourcePageMetadata {
  pageNumber: number;
  width: number;
  height: number;
}

export interface SourceBounds {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface BaseDocumentNode {
  id: string;
  sourcePage?: number;
  sourceBounds?: SourceBounds;
  sourceText?: string;
}

export interface InlineText {
  type: "text";
  text: string;
  marks?: Array<"bold" | "italic" | "underline" | "strike">;
}

export type InlineContent = InlineText;

export interface DocumentStyle {
  fontFamily?: string;
  fontSize?: number;
  fontWeight?: string | number;
  alignment?: "left" | "center" | "right" | "justify";
  lineHeight?: number;
  marginTop?: number;
  marginBottom?: number;
  indentation?: number;
}

export interface ParagraphNode extends BaseDocumentNode {
  type: "paragraph";
  content: InlineContent[];
  style: DocumentStyle;
}

export interface HeadingNode extends BaseDocumentNode {
  type: "heading";
  level: 1 | 2 | 3 | 4 | 5 | 6;
  content: InlineContent[];
  style: DocumentStyle;
}

export interface ListItemNode extends BaseDocumentNode {
  id: string;
  content: InlineContent[];
  style?: DocumentStyle;
}

export interface ListNode extends BaseDocumentNode {
  type: "bulletList" | "orderedList";
  items: ListItemNode[];
  style?: DocumentStyle;
}

export interface ColumnNode extends BaseDocumentNode {
  id: string;
  content: DocumentNode[];
  style?: DocumentStyle;
}

export interface ColumnsNode extends BaseDocumentNode {
  type: "columns";
  columns: ColumnNode[];
  style?: DocumentStyle;
}

export interface FixedLayoutNode extends BaseDocumentNode {
  type: "fixedLayout";
  content: InlineContent[];
  style?: DocumentStyle;
}

export type DocumentNode = ParagraphNode | HeadingNode | ListNode | ColumnsNode | FixedLayoutNode;

export interface StructuredDocument {
  id: string;
  sourcePdfId?: string;
  fileName: string;
  pages: SourcePageMetadata[];
  content: DocumentNode[];
}

File: backend/app/models/document.py
from typing import Annotated, Literal

from pydantic import BaseModel, Field


class ElementSource(BaseModel):
    page_number: int = Field(alias="pageNumber", ge=1)
    original_text: str | None = Field(default=None, alias="originalText")
    original_x: float | None = Field(default=None, alias="originalX")
    original_y: float | None = Field(default=None, alias="originalY")
    original_width: float | None = Field(default=None, alias="originalWidth")
    original_height: float | None = Field(default=None, alias="originalHeight")
    is_new: bool = Field(default=False, alias="isNew")

    model_config = {"populate_by_name": True}


class TextElement(BaseModel):
    id: str
    type: Literal["text"] = "text"
    content: str
    x: float
    y: float
    width: float = Field(ge=0)
    height: float = Field(ge=0)
    font_size: float = Field(alias="fontSize", gt=0)
    font_family: str = Field(alias="fontFamily")
    font_weight: Literal["normal", "bold"] = Field(default="normal", alias="fontWeight")
    font_style: Literal["normal", "italic"] = Field(default="normal", alias="fontStyle")
    color: str
    alignment: Literal["left", "center", "right"] = "left"
    rotation: float = 0
    source: ElementSource

    model_config = {"populate_by_name": True}


DocumentElement = Annotated[TextElement, Field(discriminator="type")]


class PageModel(BaseModel):
    page_number: int = Field(alias="pageNumber", ge=1)
    width: float = Field(gt=0)
    height: float = Field(gt=0)
    elements: list[DocumentElement]

    model_config = {"populate_by_name": True}


class DocumentModel(BaseModel):
    source_type: Literal["pdf"] = Field(alias="sourceType")
    file_name: str = Field(alias="fileName")
    page_count: int = Field(alias="pageCount", ge=1)
    pages: list[PageModel]

    model_config = {"populate_by_name": True}
